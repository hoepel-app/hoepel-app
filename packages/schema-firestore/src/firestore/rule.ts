export type Operation = 'read' | 'create' | 'update' | 'delete'

/**
 * Possible values:
 * - true:   Operation is always allowed
 * - false:  Operation is never allowed
 * - string: Operation is allowed if provided valid Firestore rules syntax returns true
 */
export type AllowValue = string | boolean

/** Represents a Firestore rule block (`match /.../... { ... }`) */
export class Rule {
  /**
   * @param options.comment Comment without forward slashes to add before rule
   * @param options.match Match statement, e.g. '/databases/{database}/documents/discounts/{document}'
   * @param options.extra Optional extra string to include in statement, e.g. a function
   * @param options.allow{Write,Read,Update,Delete} If (operation)s are allowed. See AllowValue for explanation
   */
  constructor(
    private readonly options: {
      readonly comment: string
      readonly match: string
      readonly extra?: string
      readonly allowCreate: AllowValue
      readonly allowRead: AllowValue
      readonly allowUpdate: AllowValue
      readonly allowDelete: AllowValue
    }
  ) {}

  private get createExpression(): string | null {
    return this.buildExpression('create', this.options.allowCreate)
  }

  private get readExpression(): string | null {
    return this.buildExpression('read', this.options.allowRead)
  }

  private get updateExpression(): string | null {
    return this.buildExpression('update', this.options.allowUpdate)
  }

  private get deleteExpression(): string | null {
    return this.buildExpression('delete', this.options.allowDelete)
  }

  private get extra(): string {
    if (this.options.extra) {
      return this.options.extra
        .split('\n')
        .map((line) => `  ${line}`)
        .join('\n')
    } else {
      return ''
    }
  }

  private buildExpression(operation: Operation, value: AllowValue): string {
    if (value === false || value === true) {
      return `allow ${operation.padEnd(6)}: if ${value.toString()};`
    } else {
      return `allow ${operation.padEnd(6)}: if ${value};`
    }
  }

  toString(): string {
    return [
      `// ${this.options.comment}`,
      `match ${this.options.match} {`,
      this.extra,
      `  ${this.createExpression}`,
      `  ${this.readExpression}`,
      `  ${this.updateExpression}`,
      `  ${this.deleteExpression}`,
      '}',
    ]
      .filter((line) => line.trim() !== '')
      .join('\n')
  }
}
