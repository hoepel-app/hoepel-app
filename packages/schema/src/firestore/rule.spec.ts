import { Rule } from './rule'

describe('Rule', () => {
    it('rule with all false', () => {
        expect(new Rule({
            comment: 'This is my comment',
            match: '/databases/{database}/documents/my-collection/{document}',
            allowRead: false,
            allowCreate: false,
            allowDelete: false,
            allowUpdate: false,
        }).toString()).toMatchSnapshot()
    })

    it('rule with all true', () => {
        expect(new Rule({
            comment: 'This is my comment',
            match: '/databases/{database}/documents/my-collection/{document}',
            allowRead: true,
            allowCreate: true,
            allowDelete: true,
            allowUpdate: true,
        }).toString()).toMatchSnapshot()
    })
    
    it('rule with permissions', () => {
        expect(new Rule({
            comment: 'This is my comment',
            match: '/databases/{database}/documents/my-collection/{document}',
            allowRead: 'request.auth.isAdmin',
            allowCreate: 'request.auth.otherThing',
            allowDelete: true,
            allowUpdate: false,
        }).toString()).toMatchSnapshot()
    })
})
