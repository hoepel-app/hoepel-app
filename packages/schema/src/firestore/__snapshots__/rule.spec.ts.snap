// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rule rule with all false 1`] = `
"// This is my comment
match /databases/{database}/documents/my-collection/{document} {
  allow create: if false;
  allow read  : if false;
  allow update: if false;
  allow delete: if false;
}"
`;

exports[`Rule rule with all true 1`] = `
"// This is my comment
match /databases/{database}/documents/my-collection/{document} {
  allow create: if true;
  allow read  : if true;
  allow update: if true;
  allow delete: if true;
}"
`;

exports[`Rule rule with permissions 1`] = `
"// This is my comment
match /databases/{database}/documents/my-collection/{document} {
  allow create: if request.auth.otherThing;
  allow read  : if request.auth.isAdmin;
  allow update: if false;
  allow delete: if true;
}"
`;
